kore/src/Kore/Unification/UnifierImpl.hs

```diff
@@ -71,123 +71,160 @@ simplifyAnds
        )
-    => UnifiedVariable variable
-    -> NonEmpty (TermLike variable)
+    => (UnifiedVariable variable, NonEmpty (TermLike variable))
     -> unifier (Predicate variable)
-solveGroupedSubstitution var patterns = do
-    predSubst <- simplifyAnds patterns
-    return Conditional
-        { term = ()
-        , predicate = Pattern.predicate predSubst
-        , substitution = Substitution.wrap $ termAndSubstitution predSubst
-        }
-  where
-    termAndSubstitution s =
-        (var, Pattern.term s) : Substitution.unwrap (Pattern.substitution s)
-
--- |Takes a potentially non-normalized substitution,
--- and if it contains multiple assignments to the same variable,
--- it solves all such assignments.
--- As new assignments may be produced during the solving process,
--- `normalizeSubstitutionDuplication` recursively calls itself until it
--- stabilizes.
-normalizeSubstitutionDuplication
+deduplicateOnce (variable, patterns) = do
+    (termLike, simplified) <- Pattern.splitTerm <$> simplifyAnds patterns
+    let substitution' = Predicate.fromSingleSubstitution (variable, termLike)
+    return (substitution' <> simplified)
+
+{- | Simplify a substitution so that each variable occurs only once.
+
+Simplify a conjunction of substitutions,
+
+@
+x₁ = t₁ ∧ ... ∧ xₙ = tₙ
+@
+
+where some of the @xᵢ@ may be the same so that each variable occurs on the
+left-hand side of only one substitution clause. New substitutions may be
+produced during simplification; @deduplicateSubstitution@ recurses until the
+solution stabilizes.
+
+See also: 'deduplicateOnce'
+
+ -}
+deduplicateSubstitution
     :: forall variable unifier
     .   ( SimplifierVariable variable
         , MonadUnify unifier
         , WithLog LogMessage unifier
         )
     => Substitution variable
     -> unifier (Predicate variable)
-normalizeSubstitutionDuplication subst
-  | null nonSingletonSubstitutions || Substitution.isNormalized subst =
-    return (Predicate.fromSubstitution subst)
-  | otherwise = do
-    predSubst <-
-        mergePredicateList
-        <$> mapM (uncurry solveGroupedSubstitution) varAndSubstList
-    finalSubst <-
-        normalizeSubstitutionDuplication
-            (  Substitution.wrap (concat singletonSubstitutions)
-            <> Conditional.substitution predSubst
-            )
+deduplicateSubstitution =
+    worker . Predicate.fromSubstitution
+  where
+    isSingleton (_, _ :| duplicates) = null duplicates
+
+    insertSubstitution
+        :: forall variable1 term
+        .  Ord variable1
+        => Map variable1 (NonEmpty term)
+        -> (variable1, term)
+        -> Map variable1 (NonEmpty term)
+    insertSubstitution multiMap (variable, termLike) =
+        let push = (termLike :|) . maybe [] Foldable.toList
+        in Map.alter (Just . push) variable multiMap
+
+    worker conditional@Conditional { predicate, substitution }
+      | all isSingleton substitutions || Substitution.isNormalized substitution
```

Comment:
Doesn't `isNormalized` imply `all isSingleton`?

Reply:
`isNormalized` does not check the substitution itself, it only checks for the `NormalizedSubstitution` constructor. This is the short-circuit for when we try to normalize a substitution which is already normalized. (To think of it, that check should be first to short-circuit properly.)